Cython/Build/IpythonMagic.py

```diff
@@ -106,6 +108,43 @@ def encode_fs(name):
         return name
 
 
+def get_encoding_candidates():
+    candidates = [sys.getdefaultencoding()]
+    try:
+        candidates.extend([sys.stdout.encoding, sys.stdin.encoding])
+    except AttributeError:
+        pass
+    try:
+        candidates.extend([sys.__stdout__.encoding, sys.__stdin__.encoding])
+    except AttributeError:
+        pass
+    unique_candidates = OrderedDict.fromkeys(candidates)
+    # encoding might be None (i.e. somebody redirects stdout):
+    unique_candidates.pop(None, None)
+    return list(unique_candidates)
```

Comment:
What about this (I'm not sure about the order of the streams yet):
```suggestion
    for stream in (sys,stdout, sys.stdin, sys.__stdout__, sys.__stdin__):
        encoding = getattr(stream, 'encoding', None)
        if encoding is not None and encoding not in candidates:
            candidates.append(encoding)
    return candidates
```
There's probably only going to be one, or sometimes two or three, different encodings here. No need to go for something algorithmically complex.

Reply:
D'accord, https://github.com/cython/cython/pull/3872/commits/8e071b06a2aa5811a43bedc22d298f0ed1f8e5ac